feat!: add basic lib and cli

DefinitelyNotSimon13 opened the pull request.

DefinitelyNotSimon13 committed:
feat: add metadata struct

DefinitelyNotSimon13 committed:
feat: add file system handler

DefinitelyNotSimon13 committed:
chore: add binary for testing/debugging

DefinitelyNotSimon13 committed:
refactor: add files module and corresponding feature

DosKobold committed:
feat: add basic structure of cli

DosKobold committed:
feat: add more flags for create command

DosKobold commented:
Not finished!

DefinitelyNotSimon13 commented on episkos_cli/src/cli.rs:
```diff
@@ -19,12 +19,39 @@ pub enum Commands {
+        non_interactive: Option<bool>,
+        /// Id of the project
+        #[arg(short, long)]
+        id: Option<String>,
```
ID is generated automatically and shouldn’t be flag.

DosKobold replied:
Done

DefinitelyNotSimon13 commented on episkos_cli/src/cli.rs:
```diff
@@ -19,12 +19,39 @@ pub enum Commands {
+        id: Option<String>,
+        /// Directory of the project
+        #[arg(short, long)]
+        directory: Option<Utf8PathBuf>,
```
not sure how you want to implemented it, but just to note it, I think that if not given directory should default to $PWD

DosKobold replied:
I think it is safer to force the user to use `--directory .` Or it would lead to unpredictable behavior (in the eyes of the user) because you have to study our documentation and may not recognize your fault.

DefinitelyNotSimon13 replied:
Fair enough, but I mean e.g. git init or cargo init both don't require that either.

DosKobold replied:
These commands are normally used by human users?! I expect the ci-script to not jump between some project directories.

DefinitelyNotSimon13 commented on episkos_cli/src/cli.rs:
```diff
@@ -19,12 +19,39 @@ pub enum Commands {
+        directory: Option<Utf8PathBuf>,
+        /// Title of the project
+        #[arg(short, long)]
+        title: Option<String>,
```
I think that you can just put String here, and when title is not given, clap automatically tells the user to use it properly?

DosKobold replied:
The default mode (without any flags) is the interactive mode. There is no need to force the user to pass the title as argument with the "title" flag. He will be asked.

DefinitelyNotSimon13 replied:
Oh okay, so basically if no title given => interactive mode?!
Just gotta think about how to handle the user passing other flags, but forgetting the title! If not in a ci environment it could be entered interactively, however that might leed to problems when in a ci env.

Maybe add like a "fail fast" flag for that case? So that the Programm just returns 1 (quits with error) if anything important is missing?

DosKobold replied:
In case the `non-interactive` flag is set it will be checked if all necessary data is given. When there is data missing, it will exit with a failure code.

DefinitelyNotSimon13 replied:
Ahh, perfect!

DefinitelyNotSimon13 commented on episkos_cli/src/main.rs:
```diff
@@ -1,25 +1,24 @@
     match &args.command {
-        cli::Commands::Create { data } => {
-            lib::create(data.clone());
+        cli::Commands::Create { non_interactive, id, directory, title, categories, languages, preferred_ide, build_systems, description , repository_url } => {
```
Idk if or how, but maybe there is a better way to so this? Idk just *feels* wrong to have so many values destructured at once

DosKobold replied:
Done with struct

DefinitelyNotSimon13 commented:
How about when run as non interactive the directory has to be given (in ci
environment) but when run as interactive it doesn’t?
And yes I just realised that doesn’t make sense, cause when interactive you
don’t pass anything anyway lol

On Thu 23. Jan 2025 at 21:49, Paul Stöckle ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In episkos_cli/src/cli.rs
> <https://github.com/SoftwareEngineeringOne/episkos/pull/6#discussion_r1927660508>
> :
>
> > @@ -19,12 +19,39 @@ pub enum Commands {
>      Create {
>          /// Create the file with given data without the interactive mode
>          #[arg(short, long)]
> -        data: Option<String>,
> +        non_interactive: Option<bool>,
> +        /// Id of the project
> +        #[arg(short, long)]
> +        id: Option<String>,
> +        /// Directory of the project
> +        #[arg(short, long)]
> +        directory: Option<Utf8PathBuf>,
>
> These commands are normally used by human users?! I expect the ci-script
> to not jump between some project directories.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/SoftwareEngineeringOne/episkos/pull/6#discussion_r1927660508>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AX5ZW4UK4QJM3RCZASDDHQT2MFIUVAVCNFSM6AAAAABVW2ZPPGVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDKNZQHE3TQNRQGA>
> .
> You are receiving this because you modified the open/close state.Message
> ID: ***@***.***>
>

DefinitelyNotSimon13 committed:
refactor: change structure for improved flexibility with database

DefinitelyNotSimon13 committed:
test: add tests for metadata builder

DosKobold committed:
feat: remove id flag

DosKobold commented on episkos_lib/src/metadata.rs:
```diff
@@ -0,0 +1,84 @@
+        MetadataBuilder::new()
+    }
+
+    pub fn update(self) -> MetadataBuilder {
```
Think there is something wrong. Cannot use `update()` more then once. The object moves away.

DosKobold replied:
There are also no tests for this component!

DefinitelyNotSimon13 committed:
fix: change id's to uuids

DefinitelyNotSimon13 committed:
fix: update code smells

DefinitelyNotSimon13 committed:
feat: add hashing for properties

DefinitelyNotSimon13 committed:
fix: dont change category name to lowercase on creation

DefinitelyNotSimon13 committed:
Merge branch 'alpha' into feat/6-3-cli-interface

DefinitelyNotSimon13 committed:
feat: add hash/checksum for metadata

DefinitelyNotSimon13 committed:
refactor!: add dedicated constructors instead of using Option

DefinitelyNotSimon13 committed:
docs: start adding documentation

DosKobold committed:
test: remove me

DefinitelyNotSimon13 committed:
docs: add more documentation for metadata and properties

DosKobold committed:
feat: implement create functionality

DosKobold committed:
Merge branch 'temporary' of https://github.com/SoftwareEngineeringOne/episkos into feat/6-3-cli-interface

DosKobold committed:
Merge branch 'feat/6-3-cli-interface' of https://github.com/SoftwareEngineeringOne/episkos into feat/6-3-cli-interface

DosKobold committed:
fix: fix vectors of flags

DosKobold committed:
feat: implement functionality for remove, add, validate

DosKobold committed:
style: change variable name

DosKobold committed:
fix: fix version handling of vector inputs

DosKobold committed:
style: format

DefinitelyNotSimon13 committed:
refactor!: rework cli

- add helper functions for conversions in lib
- differentiate between interactive/non-interactive
- improve prompts

DefinitelyNotSimon13 committed:
refactor: add modules to seperate concerns

DefinitelyNotSimon13 committed:
Merge branch 'feat/6-3-cli-interface' of https://github.com/SoftwareEngineeringOne/episkos into feat/6-3-cli-interface

DefinitelyNotSimon13 committed:
feat: add validation and removal

DefinitelyNotSimon13 committed:
docs: improve documentation

DefinitelyNotSimon13 committed:
fix: change behaviour when flags are passed in interactive mode

properties for which flags have been passed will no longer be prompted
for in interactive mode

DosKobold committed:
docs: add comments

DosKobold committed:
style: replace numbers with one constant

DosKobold committed:
style: change variable names

DefinitelyNotSimon13 committed:
refactor!: merge rename

DefinitelyNotSimon13 committed:
refactor!: rename episkos to episko as it should be
